crates/apollo-compiler/src/validation/directive.rs

```diff
@@ -16,36 +16,16 @@ pub fn check(db: &dyn ValidationDatabase) -> Vec<ApolloDiagnostic> {
     for dir_def in db.directive_definitions().iter() {
         let name = dir_def.name();
         if let Some(prev_def) = seen.get(&name) {
-            if prev_def.ast_node(db.upcast()).is_some() && dir_def.ast_node(db.upcast()).is_some() {
-                let prev_offset: usize = prev_def
-                    .ast_node(db.upcast())
-                    .unwrap()
-                    .text_range()
-                    .start()
-                    .into();
-                let prev_node_len: usize = prev_def
-                    .ast_node(db.upcast())
-                    .unwrap()
-                    .text_range()
-                    .len()
-                    .into();
+            if prev_def.loc.is_some() && dir_def.loc.is_some() {
+                let prev_offset: usize = prev_def.loc().unwrap().offset();
+                let prev_node_len: usize = prev_def.loc().unwrap().node_len();
 
-                let current_offset: usize = dir_def
-                    .ast_node(db.upcast())
-                    .unwrap()
-                    .text_range()
-                    .start()
-                    .into();
-                let current_node_len: usize = dir_def
-                    .ast_node(db.upcast())
-                    .unwrap()
-                    .text_range()
-                    .len()
-                    .into();
+                let current_offset: usize = dir_def.loc().unwrap().offset();
+                let current_node_len: usize = dir_def.loc().unwrap().node_len();
                 errors.push(ApolloDiagnostic::UniqueDefinition(UniqueDefinition {
                     ty: "directive".into(),
                     name: name.into(),
-                    src: db.input(),
+                    src: db.source_code(prev_def.loc().unwrap().file_id()),
                     original_definition: (prev_offset, prev_node_len).into(),
                     redefined_definition: (current_offset, current_node_len).into(),
```

Comment:
Here (and multiple other places) we create a diagnostic that contains two `miette::SourceSpan` which are byte ranges into a `miette::SourceCode` string stored next two them. But with multi-input the two span now could be coming from different sources.

I looked at bit at Miette’s docs and I don’t know if it’s possible to associate  two `SourceCode` strings to a single diagnostic.

Reply:
It looks like this limitation is by design :( https://github.com/zkat/miette/issues/193

Reply:
yes I noticed this too. I was hoping to be able to address it later. miette doesn't support multi-file diagnostics but you can add multiple diagnostics across different files and mark them as related, so this could be an option.

(we could also propose a re-addition to miette if we can come up with something simple enough. all we really would need is a way to pass some value T (FileId) into a SourceSpan and get that propagated to the SourceCode trait, and the rest could be implemented in apollo-rs.)